### ISSUE TITLE
Selector Execution Store

### ISSUE THREAD

**Issue by hannahhoward** [ipld/go-ipld-prime#395]
I'm pretty not sure this would live in this repo, but I'd like to explore some kind of data storage that quickly maps:

```
CID+Selector : ordered list of CIDs traversed in executing that selector
```

The simplistic version of this store could be written by executing the selector and writing a list of CIDs as you go, then being able to read back the results for that CID+selector -- a simple key value cache.

A future version could actually attempt to understand something semantic about what was being written, based on paths (and maybe selectors too), to ideally support queries for CID+selector combos that hadn't yet been written directly (not sure how far this can go, but certainly recursive all is relatively transferrable to all sub leafs)

The initial use case I have in mind is optimizing Graphsync Metadata requests (requests that only send information about CIDs in a traversal without sending blocks -- not currently implemented). You can see the ticket documented there -- https://github.com/ipfs/go-graphsync/issues/375 .

However, there are many other potential use cases.... like doing read ahead for local selector traversals to make them faster for example.

I'm not really asking for someone from IPLD to go and write this (unless someone wants to)-- I can certainly hack together a thing for Graphsync's use case, but I'd like to get feedback from IPLD stewards so I build something that makes sense.

So basically this is an early request for: how would you build such a store? What would you want the interface to look like?

**Comment by rvagg** [comment]
Closing this and redirecting folks to https://github.com/ipld/go-selector-store for continued discussion on an _actual_ implementation
